### DISCUSSION TITLE
Shared config for Android

### DISCUSSION THREAD

**Discussion by yschimke** [renovatebot/renovate#15285] [Suggest an Idea]
Curious if it's worth creating a shared config for Android Gradle projects. https://github.com/renovatebot/renovate/tree/main/lib/config/presets

```
{
  "extends": [
    "config:base"
  ],
  "packageRules": [
    {
      "groupName": "Compose Dependencies",
      "matchPackagePatterns": "^androidx\\..*compose\\."
    },
    {
      "groupName": "Kotlin Dependencies",
      "matchPackagePrefixes": ["org.jetbrains.kotlin"]
    }
  ]
}
```

Does anyone have reusable config they use on multiple projects?

**Comment by rarkins** [comment]
Do either of these have synchronised versions or are they all different?

**Reply by yschimke** [reply]
This would have synchronized versions

```
      "matchPackagePatterns": "^androidx\\.compose\\."
```

The reason for the wildcard in the middle is that you can't merge this without also merging androidx.wear.compose, but those versions differ, but do get bump at same time.

Currently they come in as many update PRs, my original reduced it to 2 (I was hoping 1).  But you really have to land it together.

**Reply by yschimke** [reply]
I believe this is one

      "matchPackagePrefixes": ["org.jetbrains.kotlin"]
      
But if I'm wrong for these, it's another good reason to have a shared config so we can get it right.

**Reply by rarkins** [reply]
If they usually must be upgraded together to work then assign them to our existing monorepo presets makes sense. Otherwise if it's grouping for convenience rather than essential then it would be best as it's own, optional preset.

**Reply by yschimke** [reply]
Could you explain the monorepo presets?

**Reply by rarkins** [reply]
They are presets of packages which need to be grouped/updated together because they are published from the same origin "monorepo"

**Reply by yschimke** [reply]
Looking at two examples, the first

https://github.com/android/wear-os-samples/pull/308 Update plugin org-jetbrains-kotlin-kapt to v1.6.21
https://github.com/android/wear-os-samples/pull/307 Update plugin org-jetbrains-kotlin-jvm to v1.6.21

I guess this can be a monorepo based on the sourceurl

```
            "depName": "org.jetbrains.kotlin:kotlin-annotation-processing-gradle",
            "groupName": "org.jetbrains.kotlin",
            ...
            "sourceUrl": "https://github.com/JetBrains/kotlin",
            "homepage": "https://kotlinlang.org/",
            "currentVersion": "1.6.10",
```

**Reply by yschimke** [reply]
But for the compose ones, it doesn't look like that will work.

Too many things use https://cs.android.com/androidx/platform/frameworks/support

```
          {
            "depName": "androidx.compose.compiler:compiler",
            "groupName": "androidx.compose.compiler",
            "currentValue": "1.2.0-alpha07",
            "managerData": {
              "fileReplacePosition": 1702,
              "packageFile": "gradle/libs.versions.toml"
            },
            "fileReplacePosition": 1702,
            "registryUrls": [
              "https://repo.maven.apache.org/maven2",
              "https://plugins.gradle.org/m2/",
              "https://dl.google.com/android/maven2/",
              "https://oss.sonatype.org/content/repositories/snapshots"
            ],
            "depIndex": 14,
            "warnings": [],
            "versioning": "gradle",
            "sourceUrl": "https://cs.android.com/androidx/platform/frameworks/support",
            "homepage": "https://developer.android.com/jetpack/androidx/releases/compose-compiler#1.1.1",
            "currentVersion": "1.2.0-alpha07",
            "isSingleVersion": true,
            "fixedVersion": "1.2.0-alpha07",
```

**Comment by yschimke** [comment]
Another tangential topic, libraries like compose, or compose-compiler are tied to kotlin versions

https://developer.android.com/jetpack/androidx/releases/compose-kotlin

Without this mapping, renovate launches repeated PRs that can't work

https://github.com/android/wear-os-samples/pull/467
